vm/gc.go

```diff
@@ -0,0 +1,112 @@
+package vm
+
+import (
+	"time"
+
+	"github.com/jbenet/goprocess"
+	"github.com/libp2p/go-smart-record/ir"
+)
+
+func (v *vm) gcLoop(proc goprocess.Process) {
+	msgSyncTicker := time.NewTicker(v.gcPeriod)
+	defer msgSyncTicker.Stop()
+	for {
+		select {
+		case <-msgSyncTicker.C:
+			// NOTE: Locking all keys while garbage collecting may really
+			// harm performance, specially if the gcPeriod is low. Consider
+			// adding a lock per entry or other schemes to improve this.
+			// It'd be useful to gather some metrics.
+			v.lk.Lock()
+			v.garbageCollect()
+			v.lk.Unlock()
+
+		case <-proc.Closing():
+			return
+		}
+	}
+}
+
+func (v *vm) garbageCollect() {
+	// For each record
+	for _, r := range v.keys {
+		// And the datastore of each peer
+		for p, entry := range *r {
+			e, gc := gcDict(*entry)
```

Comment:
So, here you are suffering from having IR nodes be values (as opposed to pointers).
Each time garbage collection runs, it effectively makes a fresh new copy of the entire record
so it can return it as the result. This is very inefficient.

Reply:
Totally agree! This is fixed no longer a problem after making IR nodes immutable.